localstack/services/firehose/provider.py

```diff
@@ -0,0 +1,610 @@
+                for record in records:
+                    obj_id = uuid.uuid4()
+
+                    data = "{}"
+                    # DirectPut
+                    if "Data" in record:
+                        data = base64.b64decode(record["Data"])
+                    # KinesisAsSource
+                    elif "data" in record:
+                        data = base64.b64decode(record["data"])
+
+                    try:
+                        body = json.loads(data)
+                    except Exception as e:
+                        LOG.warning("Elasticsearch only allows json input data!")
+                        raise e
+
+                    LOG.debug(
+                        "Publishing to ES destination. Data: %s", truncate(data, max_length=300)
+                    )
+                    try:
+                        es.create(index=es_index, doc_type=es_type, id=obj_id, body=body)
+                    except Exception as e:
+                        LOG.exception(f"Unable to put record to stream {delivery_stream_name}.")
+                        raise e
+            if "AmazonopensearchserviceDestinationDescription" in destination:
+                # TODO implement opensearch integration
+                # - Replace the elasticsearch client with the opensearch client
+                # - Then this most likely can be handled together with ElasticsearchDestinationDescription
+                # opensearch_dest = destination["AmazonopensearchserviceDestinationDescription"]
+                pass
+            if "S3DestinationDescription" in destination:
+                s3_dest_desc = destination["S3DestinationDescription"]
+                self._put_records_to_s3_bucket(delivery_stream_name, records, s3_dest_desc)
+            if "HttpEndpointDestinationDescription" in destination:
+                http_dest = destination["HttpEndpointDestinationDescription"]
+                end_point = http_dest["EndpointConfiguration"]
+                url = end_point["Url"]
+                record_to_send = {
+                    "requestId": str(uuid.uuid4()),
+                    "timestamp": (int(time.time())),
+                    "records": [],
+                }
+                for record in records:
+                    data = record.get("Data") or record.get("data")
+                    record_to_send["records"].append({"data": to_str(data)})
+                headers = {
+                    "Content-Type": "application/json",
+                }
+                try:
+                    requests.post(url, json=record_to_send, headers=headers)
+                except Exception as e:
+                    LOG.exception(f"Unable to put Firehose records to HTTP endpoint {url}.")
+                    raise e
+        return [
+            PutRecordBatchResponseEntry(RecordId=str(uuid.uuid4())) for _ in unprocessed_records
+        ]
+
+    def _add_missing_record_attributes(self, records: List[Dict]) -> None:
```

Comment:
is there a specific case you have in mind where these attributes would be missing or is this just for general safety/robustness?

Reply:
The method is a bit generically named, but essentially it currently only adds Kinesis metadata if it's missing.